File: <100/070/070.hs
import Math.NumberTheory.ArithmeticFunctions ( totient )
import Data.List ( elemIndex, sort )


main :: IO ()
main = print $ (+2) <$> minimumAt totientRates


totientRates :: [Double]
totientRates = map totientRate [2..upperLimit]


totientRate :: Int -> Double
totientRate n
    | isOK      = nF / phi
    | otherwise = fromIntegral upperLimit
    where
        phi' = totient n
        phi  = fromIntegral phi'
        nF   = fromIntegral n
        isOK = s == t
        s    = sort . show $ n
        t    = sort . show $ phi'


minimumAt :: Ord a => [a] -> Maybe Int
minimumAt list = elemIndex ( minimum list ) list


upperLimit :: Int
upperLimit = 10^7-1

File: <100/070/070.jl
using Primes


stringToArray(s:: String):: Array{Char} = [c for c ∈ s]


function arePermutes(n:: Int, m:: Int):: Bool
  s = n |> string |> stringToArray |> sort
  t = m |> string |> stringToArray |> sort
  return s == t
end


function totientRate(n:: Int):: Float64
  phi = totient(n)
  if arePermutes(n, phi)
    return n / phi
  else
    return 10^7
  end
end


function main():: Nothing
  upper_limit = 10^7-1
  info_min = map(totientRate, 2:upper_limit) |> findmin
  info_min[2] + 1 |> println
  return nothing 
end


@time main()
